Sources/SwiftSyntax/Utils.swift

```diff
@@ -10,62 +10,86 @@
 //
 //===----------------------------------------------------------------------===//
 
-public struct ByteSourceRange: Equatable, Sendable {
-  public var offset: Int
-  public var length: Int
+@available(*, deprecated, message: "Use Range<AbsolutePosition> instead")
+public typealias ByteSourceRange = Range<AbsolutePosition>
 
-  public init(offset: Int, length: Int) {
-    self.offset = offset
-    self.length = length
+public extension Range<AbsolutePosition> {
+  @available(*, deprecated, message: "Use lowerBound.utf8Offset instead")
+  var offset: Int { lowerBound.utf8Offset }
+
+  @available(*, deprecated, message: "Construct a Range<AbsolutePosition> instead")
+  init(offset: Int, length: Int) {
+    self = AbsolutePosition(utf8Offset: offset)..<AbsolutePosition(utf8Offset: offset + length)
   }
 
-  public var endOffset: Int {
-    return offset + length
+  @available(*, deprecated, message: "Use upperBound.utf8Offset instead")
+  var endOffset: Int {
+    return upperBound.utf8Offset
   }
 
-  public var isEmpty: Bool {
-    return length == 0
+  /// Returns the range for the overlapping region between two ranges.
+  ///
+  /// If the intersection is empty, this returns a range starting at offset 0 and having length 0.
+  func intersected(_ other: Range<AbsolutePosition>) -> Range<AbsolutePosition> {
+    return self.intersecting(other) ?? AbsolutePosition(utf8Offset: 0)..<AbsolutePosition(utf8Offset: 0)
+  }
+}
+
+public extension Range<AbsolutePosition> {
+  /// The number of bytes between the ranges lower bound and its upper bound
+  var length: SourceLength { return SourceLength(utf8Length: upperBound.utf8Offset - lowerBound.utf8Offset) }
+
+  init(position: AbsolutePosition, length: SourceLength) {
+    self = position..<(position + length)
   }
 
-  public func intersectsOrTouches(_ other: ByteSourceRange) -> Bool {
-    return self.endOffset >= other.offset && self.offset <= other.endOffset
+  /// Returns `true` if they intersection between this range and `other` is non-empty or if the two ranges are directly
+  /// adjacent to each other.
+  func intersectsOrTouches(_ other: Range<AbsolutePosition>) -> Bool {
+    return self.upperBound >= other.lowerBound && self.lowerBound <= other.upperBound
   }
 
-  public func intersects(_ other: ByteSourceRange) -> Bool {
-    return self.endOffset > other.offset && self.offset < other.endOffset
+  /// Returns `true` if the intersection between this range and `other` is non-empty.
+  func intersects(_ other: Range<AbsolutePosition>) -> Bool {
```

Comment:
I couldn't agree more! Adding an access control attribute to an extension makes code much less readable. I believe this could be handled by `swift-format`, but from what I see in the documentation, there is no rule for that right now. 

It seems to be possible to implement such a rule, doesn't it, @allevato?

Reply:
Here is a draft PR with this rule turned on: #2602.

Reply:
I think I even commented somewhere this week that I dislike `public extension` myself. Changed it.

Reply:
Yeah, I believe we've all said that... we just don't have the check for it :)

Reply:
No one likes them... :D